data_structures/reverse_a_linked_list.cpp

```diff
@@ -0,0 +1,181 @@
+ */
+class list {
+ private:
+    Node *head;  // link before the actual first element
+ public:
+    /**
+     * List constructor. Initializes the first link.
+     */
+    list() {
+        head = nullptr;  // Initialize the first link
+    }
+    bool isEmpty();
+    void insert(int32_t new_elem);
+    void reverseList();
+    void display();
+    int32_t top();
+    int32_t last();
+};
+
+/**
+ * @brief Utility function that checks if the list is empty
+ * @returns true if the list is empty
+ * @returns false if the list is not empty
+ */
+bool list::isEmpty() { return head == nullptr; }
+
+/**
+ * @brief Utility function that adds a new element at the end of the list
+ * @param new_elem element be added at the end of the list
+ */
+void list::insert(int32_t n) {
+    try {
+        Node *new_node = new Node();
+        Node *temp = nullptr;
+        new_node->val = n;
+        new_node->next = nullptr;
+        if (isEmpty()) {
+            head = new_node;
+        } else {
+            temp = head;
+            while (temp->next != nullptr) {
+                temp = temp->next;
+            }
+            temp->next = new_node;
+        }
+    } catch (std::bad_alloc &exception) {
+        std::cerr << "bad_alloc detected: " << exception.what() << "\n";
+    }
+}
+
+/**
+ * @brief Utility function for reversing a list
+ * @brief Using the current, previous, and next pointer.
+ * @returns void
+ */
+void list::reverseList() {
+    Node *curr = head;
+    Node *prev = nullptr, *next_node = nullptr;
+    while (curr != nullptr) {
+        next_node = curr->next;
+        curr->next = prev;
+        prev = curr;
+        curr = next_node;
+    }
+    head = prev;
+}
+
+/**
+ * @brief Utility function to find the top element of the list
+ * @returns the top element of the list
+ */
+int32_t list::top() {
+    try {
+        if (!isEmpty()) {
+            return head->val;
+        }
+    } catch (const std::exception &e) {
+        std::cerr << "List is empty" << e.what() << '\n';
+    }
+}
+/**
+ *  @brief Utility function to find the last element of the list
+ *  @returns the last element of the list
+ */
+int32_t list::last() {
+    try {
+        if (!isEmpty()) {
+            Node *t = head;
+            while (t->next != nullptr) {
+                t = t->next;
+            }
+            return t->val;
+        }
+    } catch (const std::exception &e) {
+        std::cerr << "List is empty" << e.what() << '\n';
+    }
+}
+}  // namespace linked_list
+}  // namespace data_structures
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    data_structures::linked_list::list L;
+    // 1st test
+    L.insert(11);
+    L.insert(12);
+    L.insert(15);
+    L.insert(10);
+    L.insert(-12);
+    L.insert(-20);
+    L.insert(18);
+    assert(L.top() == 11);
+    assert(L.last() == 18);
+    L.reverseList();
+    // Reversal Testing
+    assert(L.top() == 18);
+    assert(L.last() == 11);
```

Comment:
Just testing the first and last elements is not enough. Try to frame proper tests. What if anything in between goes wrong? These tests would not be able to detect such error.